### PR TITLE
Remote log

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add a remote log capability into HELICS

### Proposed changes

 - add remote log commands
 - add remote log support in the logging system
 -  move the logging operations to a dedicated class

### CODE REVIEW

**Comment by phlptp** [comment]
this is waiting on the processComms PR to make sure some tests are reliable for this PR

**Review by trevorhardy — APPROVED** [review]
Based on the documentation, it doesn't look like it but I want to check: anything need to be added to the Configuration Options Reference?

**Comment by phlptp** [comment]
All the functionality is run through the command interface.  No way to specify via command line hence nothing in the config options reference.

File: src/helics/common/logging.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    NO_PRINT = HELICS_LOG_LEVEL_NO_PRINT,  //!< never print
+    ERROR_LEVEL = HELICS_LOG_LEVEL_ERROR,  //!< only print errors
+    PROFILING = HELICS_LOG_LEVEL_PROFILING,  //!< profiling log level
+    WARNING = HELICS_LOG_LEVEL_WARNING,  //!< print/log warning and errors
+    SUMMARY = HELICS_LOG_LEVEL_SUMMARY,  //!< print/log summary information
+    CONNECTIONS =
+        HELICS_LOG_LEVEL_CONNECTIONS,  //!< print summary+ federate level connection information
+    INTERFACES =
+        HELICS_LOG_LEVEL_INTERFACES,  //!< print connections +interface level connection information
+    TIMING = HELICS_LOG_LEVEL_TIMING,  //!< print interfaces+ timing(exec/grant/disconnect)
```
Comment (nightlark) [inline src/helics/common/logging.hpp:31]:
```suggestion
        HELICS_LOG_LEVEL_CONNECTIONS,  //!< print summary+federate level connection information
    INTERFACES =
        HELICS_LOG_LEVEL_INTERFACES,  //!< print connections+interface level connection information
    TIMING = HELICS_LOG_LEVEL_TIMING,  //!< print interfaces+timing(exec/grant/disconnect)
```

File: src/helics/core/LogManager.hpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    /// default logging object to use if the logging callback is not specified
+    std::shared_ptr<spdlog::logger> consoleLogger;
+    /// default logging object to use if the logging callback is not specified
+    std::shared_ptr<spdlog::logger> fileLogger;
+    std::atomic<bool> initialized{false};
+    mutable LogBuffer mLogBuffer;  //!< object for buffering a set of log messages
+    /** a logging function for logging or printing messages*/
+    std::function<void(int, std::string_view, std::string_view)> loggerFunction;
+    std::function<void(ActionMessage&& mm)> mTransmit;
+    std::string logFile;  //!< the file to log message to
```
Comment (nightlark) [inline src/helics/core/LogManager.hpp:54]:
```suggestion
    std::string logFile;  //!< the file to log messages to
```

File: src/helics/core/LogManager.hpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  public:
+    /// force the log to flush after every message
+    std::atomic<bool> forceLoggingFlush{false};
+
+    ~LogManager();
+
+    void initializeLogging(const std::string& identifier);
+
+    /** send a Message to the logging system
```
Comment (nightlark) [inline src/helics/core/LogManager.hpp:64]:
```suggestion
    /** send a message to the logging system
```

**Review by nightlark — APPROVED** [review]

**Pull request merged**